src/wavm/WAVMWasmModule.cpp

```diff
@@ -643,50 +682,99 @@ int WAVMWasmModule::dynamicLoadModule(const std::string& path,
 {
     // This function is essentially dlopen. See the comments around the GOT
     // function for more detail on the dynamic linking approach.
+    // It returns 0 on error, as does dlopen
 
     auto logger = faabric::util::getLogger();
 
     // Return the handle if we've already loaded this module
     if (dynamicPathToHandleMap.count(path) > 0) {
-        logger->debug("Reusing dynamic module {}", path);
-        return dynamicPathToHandleMap[path];
+        int cachedHandle = dynamicPathToHandleMap[path];
+        logger->debug(
+          "Using cached dynamic module handle {} for {}", cachedHandle, path);
+        return cachedHandle;
+    }
+
+    // Work out if we're loading an existing module or using the fallback
+    int thisHandle;
+    if (path.empty()) {
+        logger->warn("Dynamic linking main module");
+        return MAIN_MODULE_DYNLINK_HANDLE;
+    } else if (boost::filesystem::is_directory(path)) {
+        logger->warn("Dynamic linking a directory {}", path);
+        return 0;
+    } else if (!boost::filesystem::exists(path)) {
+        logger->warn("Dynamic module {} does not exist", path);
+        return 0;
```

Comment:
The previous implementation didn't conform to the `dlopen` interface, where you are allowed to pass `nullptr` to get a reference to the currently executing shared object. In our case, this currently executing shared object is just the function itself, so we use this `MAIN_MODULE_DYNLINK_HANDLE` to refer back to the main module.